### PR TITLE
[mle] generate challenge for "Child Update Request" early

### PR DESCRIPTION
**Pull request by sarveshkumarv3**
PR to address the issue discussed [here](https://github.com/openthread/openthread/issues/9787)

### CODE REVIEW

**Comment by abtink** [comment]
Thanks @sarveshkumarv3 .

I suggest adding a more detailed commit message and PR description to tell what is changed and the reason for the change. This will help folks looking at the commit history :)

Here is my suggestion:
```
[mle] generate challenge for "Child Update Request" early

This commit updates how random challenge is generated for "Child
Update Request" message sent by a parent during child restoration
(after parent reset/re-attach).

A random challenge is now generated and saved in the `Child` entry
when it is first initialized in `kRestoredState`. The
`SendChildUpdateRequest()` will use the saved challenge rather than
generating a new one.

This change prevents overwriting the saved challenge when the child is
also detached and happens to send a "Parent Request" in the window
where the parent transitions to the router/leader role and before the
parent sends the "Child Update Request". It ensures that the same
random challenge is included in both "Parent Response" and "Child
Update Response," guaranteeing proper acceptance of the
child's "Child ID request".
```

File: src/core/thread/mle_router.cpp [hunk 1]
```diff
@@ -2962,7 +2961,10 @@ Error MleRouter::SendChildUpdateRequest(Child &aChild)
     if (!aChild.IsStateValid())
     {
         SuccessOrExit(error = message->AppendTlvRequestTlv(kTlvs));
-        aChild.GenerateChallenge();
+        if (!aChild.IsStateRestored())
+        {
```
Comment (abtink) [inline src/core/thread/mle_router.cpp:2965]: May also help adding a short comment here. Folks looking a the code will wonder why the `GenerateChallenge()` is being skipped under `IsStateRestored()`.

File: src/core/thread/mle_router.cpp [hunk 2]
```diff
@@ -2962,7 +2961,10 @@ Error MleRouter::SendChildUpdateRequest(Child &aChild)
...
+            aChild.GenerateChallenge();
+        }
```
Comment (abtink) [inline src/core/thread/mle_router.cpp:2967]: How about this?

```suggestion

        if (!aChild.IsStateRestored())
        {
            // A random challenge is generated and saved when `aChild`
            // is first initialized in `kStateRestored`. We will use
            // the saved challenge here. This prevents overwriting
            // the saved challenge when the child is also detached
            // and happens to send a "Parent Request" in the window
            // where the parent transitions to the router/leader role
            // and before the parent sends the "Child Update Request". 
            // This ensures that the same random challenge is
            // included in both "Parent Response" and "Child Update
            // Response," guaranteeing proper acceptance of the
            // child's "Child ID request".

            aChild.GenerateChallenge();
        }

```

File: src/core/thread/mle_router.cpp (hunk 1 above)
Comment (sarveshkumarv3) [reply src/core/thread/mle_router.cpp:2966]: thanks, incorporated your comments to the code snippet and and the commit log.

**Review by jwhui — APPROVED** [review]

**Pull request merged**